### PR TITLE
write react-toastify study

### PR DESCRIPTION
**Pull request by jackbsteinberg**

### CODE REVIEW

**Comment by jackbsteinberg** [comment]
Didn't mark this in the commit message, but completed both `react-toastify` and android `toast`

File: study-group/Toast-Implementation-Studies.md [hunk 1]
```diff
@@ -70,25 +70,164 @@ The `<paper-toast>` holds onto information about how it was dismissed the last t
...
+
+Will show the toast further down and further right.
+The toast can also be injected with a custom layout, which requires more complex Kotlin, explained [here](https://developer.android.com/guide/topics/ui/notifiers/toasts#CustomToastView).
+The `text`, `view`, `gravity`, and `duration` of the toast are all the elements the developer can control when working with this very straightforward component.
+
+
+### Notable Features & Details
+
+#### Gravity
+The library takes an interesting approach to positioning, by defining letting the user define the "center of gravity" of the element, along with offset values for the axes of gravity.
```
Comment (fergald) [inline study-group/Toast-Implementation-Studies.md]:
```suggestion
The library takes an interesting approach to positioning, by letting the user define the "center of gravity" of the element, along with offset values for the axes of gravity.
```

Comment (fergald) [inline study-group/Toast-Implementation-Studies.md]: Missing semantic line breaks. I think I see the problem. You can and should break within sentences, e.g. this should probably be 3 lines.

Comment (fergald) [inline study-group/Toast-Implementation-Studies.md]: Gravity (not center of gravity) is a standard android UI concept, maybe just point that out.

File: study-group/Toast-Implementation-Studies.md [hunk 2]
```diff
@@ -70,25 +70,164 @@ The `<paper-toast>` holds onto information about how it was dismissed the last t
...
+        <button onClick={this.notify}>Notify !</button>
+        <ToastContainer />
+      </div>
+    );
+  }
+}
+```
+
+`react-toastify` uses the explicit component pattern, by having the developer render a React `ToastContainer` component (the recommendation is to put this in the application root).
+This single component will render all toasts called from JavaScript, and can be configured with global options.
```
Comment (fergald) [inline study-group/Toast-Implementation-Studies.md]: don't think you need a "," here.

File: study-group/Toast-Implementation-Studies.md [hunk 3]
```diff
@@ -70,25 +70,164 @@ The `<paper-toast>` holds onto information about how it was dismissed the last t
...
+```js
+class App extends Component {
+  tenSecondToast = () => toast("This toast will self destruct in 10...", { autoClose: 10000 });
+  normalToast = () => toast("This is a normal toast");
+
+  render(){
+    return (
+      <div>
+        <button onClick={this.tenSecondToast}>Close after 10 seconds</button>
+        <button onClick={this.normalToast}>Close after default amount of time</button>
```
Comment (fergald) [inline study-group/Toast-Implementation-Studies.md]: This example was unclear to me until I read the text below. I couldn't tell if "default" was referring to 5000 or a more global default.

I would suggest
```suggestion
        <button onClick={this.normalToast}>Close after 5000ms</button>
```

File: study-group/Toast-Implementation-Studies.md [hunk 4]
```diff
@@ -70,25 +70,164 @@ The `<paper-toast>` holds onto information about how it was dismissed the last t
...
+
+```js
+class App extends Component {
+  tenSecondToast = () => toast("This toast will self destruct in 10...", { autoClose: 10000 });
+  normalToast = () => toast("This is a normal toast");
+
+  render(){
+    return (
+      <div>
+        <button onClick={this.tenSecondToast}>Close after 10 seconds</button>
```
Comment (fergald) [inline study-group/Toast-Implementation-Studies.md]:
```suggestion
        <button onClick={this.tenSecondToast}>Close after 10000ms seconds</button>
```
I think it's easier on the reader to use the same numbers as the code with the API's units.

File: study-group/Toast-Implementation-Studies.md [hunk 5]
```diff
@@ -70,25 +70,164 @@ The `<paper-toast>` holds onto information about how it was dismissed the last t
...
+        <button onClick={this.normalToast}>Close after default amount of time</button>
+        <ToastContainer autoClose={5000} />
+      </div>
+    );
+  }
+}
+```
+
+The first button creates a toast that dismisses in 10 seconds, overriding the `ToastContainer` default.
```
Comment (fergald) [inline study-group/Toast-Implementation-Studies.md]: I'm not sure that this text is adding much value as the example also explains the outcome. Perhaps move anything extra here into the example?

File: study-group/Toast-Implementation-Studies.md [hunk 6]
```diff
@@ -70,25 +70,164 @@ The `<paper-toast>` holds onto information about how it was dismissed the last t
...
+The second provides no `autoClose` value, so it dismisses in the default 5 seconds.
+For some of the more common controls, the toast can be configured like so:
```
Comment (fergald) [inline study-group/Toast-Implementation-Studies.md]: new paragraph?

File: study-group/Toast-Implementation-Studies.md [hunk 7]
```diff
@@ -70,25 +70,164 @@ The `<paper-toast>` holds onto information about how it was dismissed the last t
...
+For instance, with the toast ID the toast can be dismissed, updated, and checked to see if it's active.
+
+#### Mobile
+On mobile toasts will automatically take all available width, and all toasts are swipeable / draggable to dismiss.
+
+#### Styling
+Library provides fields `className`, `toastClassName`, and, `progressClassName` for the developer to pass in CSS classes.
+
+#### Render Custom Component
+To customize the view the `toast` function can take in a React component, which will receive the function required to close the toast as props (though it won't get the toast ID).
```
Comment (fergald) [inline study-group/Toast-Implementation-Studies.md]: Can you rephrase this? I don't know what this means.

File: study-group/Toast-Implementation-Studies.md [hunk 8]
```diff
@@ -70,25 +70,164 @@ The `<paper-toast>` holds onto information about how it was dismissed the last t
...
+The helper allows the dev to set classes and durations for `enter` and `exit`, as well as make transitions position specific.
+
+### Takeaways
+- Extremely unopinionated and configurable; the developer has near total control.
+- Component pattern is optional, only needs one component, but allows multiple.
+- Close button is default, but can be replaced with a custom close button.
+  - Separate from passing in a custom React Component.
+- The toast can be configured to display after a certain delay.
+  - I'm not sure what the advantage this has over firing the toast in a setTimeout.
+- Passing null into a property resets it to default (useful for `update()`).
```
Comment (fergald) [inline study-group/Toast-Implementation-Studies.md:261]: What is update()? It's not referenced above.

File: study-group/Toast-Implementation-Studies.md (hunk 2 above)
Comment (jackbsteinberg) [reply study-group/Toast-Implementation-Studies.md]:
```suggestion
This single component will render all toasts called from JavaScript and can be configured with global options.
```

File: study-group/Toast-Implementation-Studies.md (hunk 3 above)
Comment (jackbsteinberg) [reply study-group/Toast-Implementation-Studies.md]:
```suggestion
        <button onClick={this.normalToast}>Close after 5000ms (default)</button>
```

File: study-group/Toast-Implementation-Studies.md (hunk 4 above)
Comment (jackbsteinberg) [reply study-group/Toast-Implementation-Studies.md]: You're right that makes sense

File: study-group/Toast-Implementation-Studies.md (hunk 8 above)
Comment (jackbsteinberg) [reply study-group/Toast-Implementation-Studies.md:261]: Was not clear enough with this, changing it now

File: study-group/Toast-Implementation-Studies.md (hunk 1 above)
Comment (jackbsteinberg) [reply study-group/Toast-Implementation-Studies.md]: I'm still getting used to this semantic line break pattern, especially when in the middle of sentences.
I'll add this in now.

File: study-group/Toast-Implementation-Studies.md (hunk 7 above)
Comment (jackbsteinberg) [reply study-group/Toast-Implementation-Studies.md]: This was unclear, rephrasing now

File: study-group/Toast-Implementation-Studies.md [hunk 9]
```diff
@@ -70,25 +70,190 @@ The `<paper-toast>` holds onto information about how it was dismissed the last t
...
+The library takes the Android approach to positioning,
+by letting the user define the Gravity of the element,
+along with offset values for the axes.
+
+#### Duration Constants
+The library comes with baked in constants for displaying the toast for long and short periods of time.
+
+### Takeaways
+- High opinion, low control.
+- The API for the `Toast` object provides getters and setters for each of `text`, `view`, `gravity`, `offset` and `duration`.
```
Comment (domenic) [inline study-group/Toast-Implementation-Studies.md:133]: Does this mean they can be changed dynamically, while the toast is displaying? I hadn't thought of that before, but I guess it makes sense. I'm not sure if it's worth mentioning explicitly, but I found it interesting.

File: study-group/Toast-Implementation-Studies.md [hunk 10]
```diff
@@ -70,25 +70,190 @@ The `<paper-toast>` holds onto information about how it was dismissed the last t
...
+- Beyond getters and setters, the only control the object allows is `.makeText()`, `.show()` and `.cancel()`.
```
Comment (domenic) [inline study-group/Toast-Implementation-Studies.md]: I'd omit `makeText()` since it's the factory method, not really about controlling.

File: study-group/Toast-Implementation-Studies.md [hunk 11]
```diff
@@ -70,25 +70,190 @@ The `<paper-toast>` holds onto information about how it was dismissed the last t
...
+
+## [react-toastify](https://github.com/fkhadra/react-toastify)
+`react-toastify` is a React component library by author [@fkhadra](https://github.com/fkhadra) for rich, extensible notification toasts.
+
+### Design & Sample Code
+
+To use a `react-toastify` toast requires some simple JavaScript:
```
Comment (domenic) [inline study-group/Toast-Implementation-Studies.md]: Avoid using "simple"; it's a judgement on the reader's ability that can feel bad if they find the code harder to understand.

Comment (domenic) [inline study-group/Toast-Implementation-Studies.md]: Maybe "react code" instead of "JavaScript" since this uses JSX.

File: study-group/Toast-Implementation-Studies.md [hunk 12]
```diff
@@ -70,25 +70,190 @@ The `<paper-toast>` holds onto information about how it was dismissed the last t
...
+
+```js
```
Comment (domenic) [inline study-group/Toast-Implementation-Studies.md]: I'm curious if using `jsx` instead of `js` give different results for syntax highlighting.

File: study-group/Toast-Implementation-Studies.md [hunk 13]
```diff
@@ -70,25 +70,190 @@ The `<paper-toast>` holds onto information about how it was dismissed the last t
...
+class App extends Component {
+  notify = () => toast("Very simple toast!");
+
+  render(){
+    return (
+      <div>
+        <button onClick={this.notify}>Notify !</button>
+        <ToastContainer />
```
Comment (domenic) [inline study-group/Toast-Implementation-Studies.md:156]: Maybe show the importing of toast/ToastContainer?

File: study-group/Toast-Implementation-Studies.md [hunk 14]
```diff
@@ -70,25 +70,190 @@ The `<paper-toast>` holds onto information about how it was dismissed the last t
...
+class App extends Component {
+  // Creating a toast that will close in 10000ms.
+  tenSecondToast = () => toast("This toast will self destruct in 10...", { autoClose: 10000 });
+  // Creating a toast that will close in 5000ms (the default).
+  normalToast = () => toast("This is a normal toast");
+
+  render(){
+    return (
+      <div>
+        <button onClick={this.tenSecondToast}>Close after 10000ms seconds</button>
```
Comment (domenic) [inline study-group/Toast-Implementation-Studies.md]:
```suggestion
        <button onClick={this.tenSecondToast}>Close after 10000ms</button>
```

File: study-group/Toast-Implementation-Studies.md [hunk 15]
```diff
@@ -70,25 +70,190 @@ The `<paper-toast>` holds onto information about how it was dismissed the last t
...
+```
+
+For some of the common controls, the toast can be configured like so:
+
+```js
+class App extends Component {
+  topRightWarning = () => toast('Warning!', {
+    type: toast.TYPE.WARNING,
+    position: toast.POSITION.TOP_RIGHT,
+    onClose: ({ props }) => window.alert('I warned you') // props: all props passed to rendered toast component
```
Comment (domenic) [inline study-group/Toast-Implementation-Studies.md]: Maybe just omit the props argument since it's not essential to the example? i.e. `() => window.alert(...)`

File: study-group/Toast-Implementation-Studies.md [hunk 16]
```diff
@@ -70,25 +70,190 @@ The `<paper-toast>` holds onto information about how it was dismissed the last t
...
+  }
+}
+```
+
+As in this example, the `toastId` gives the developer control over the toast beyond configuration and display.
+
+### Notable Features & Details
+
+#### Toast ID
+the `toast()` function returns an ID of the toast it creates, which can be used to act upon that toast.
```
Comment (domenic) [inline study-group/Toast-Implementation-Studies.md]:
```suggestion
The `toast()` function returns an ID of the toast it creates, which can be used to act upon that toast.
```

File: study-group/Toast-Implementation-Studies.md [hunk 17]
```diff
@@ -70,25 +70,190 @@ The `<paper-toast>` holds onto information about how it was dismissed the last t
...
+For instance, with the toast ID the toast can be used with `dismiss()`, `update()`, and `isActive()`.
```
Comment (domenic) [inline study-group/Toast-Implementation-Studies.md:238]: I wonder what the motivation is, for using IDs instead of JavaScript objects with methods.

**Review by domenic — APPROVED** [review]
LGTM with nits. Be sure to update the commit message to mention both studies when merging.

File: study-group/Toast-Implementation-Studies.md (hunk 9 above)
Comment (jackbsteinberg) [reply study-group/Toast-Implementation-Studies.md:133]: To be honest, I'm not sure. The way the toast itself works is that you create and configure the object, then show it. 

What I think is probably more likely is that `show()` just shows the toast with the previously configured options, and setting the options mid-display will only affect the toast the next time you call `show()`.

File: study-group/Toast-Implementation-Studies.md (hunk 10 above)
Comment (jackbsteinberg) [reply study-group/Toast-Implementation-Studies.md]: Makes sense, on it
```suggestion
- Beyond getters and setters, the only controls the object allows are `.show()` and `.cancel()`.
```

File: study-group/Toast-Implementation-Studies.md (hunk 11 above)
Comment (jackbsteinberg) [reply study-group/Toast-Implementation-Studies.md]: Completely understand, definitely doesn't need to be there
```suggestion
Using a `react-toastify` toast requires some JavaScript:
```

File: study-group/Toast-Implementation-Studies.md (hunk 12 above)
Comment (jackbsteinberg) [reply study-group/Toast-Implementation-Studies.md]: Ooh it does, that's good to know

File: study-group/Toast-Implementation-Studies.md (hunk 13 above)
Comment (jackbsteinberg) [reply study-group/Toast-Implementation-Studies.md:156]: Good call

File: study-group/Toast-Implementation-Studies.md (hunk 14 above)
Comment (jackbsteinberg) [reply study-group/Toast-Implementation-Studies.md]: Poor editing, thank you

File: study-group/Toast-Implementation-Studies.md (hunk 15 above)
Comment (jackbsteinberg) [reply study-group/Toast-Implementation-Studies.md]: I think I can find a better way to include them
```suggestion
    onClose: ({ props }) => window.alert(`You just closed the ${props.type} message`) // props: all props passed to rendered toast component
```

File: study-group/Toast-Implementation-Studies.md (hunk 17 above)
Comment (jackbsteinberg) [reply study-group/Toast-Implementation-Studies.md:238]: I'm not sure, I'm going to add a note to the takeaways to make sure that's noted
